pkg/metrics/metrics.go

```diff
@@ -0,0 +1,141 @@
+package metrics
+
+import (
+	"net/http"
+	"sync"
+
+	"k8s.io/component-base/metrics"
+	"k8s.io/klog/v2"
+)
+
+var (
+	r    *metricRecorder // singleton instance of metricRecorder
+	once sync.Once
+)
+
+type metricRecorder struct {
+	registry metrics.KubeRegistry
+	metrics  map[string]interface{}
+}
+
+// Recorder returns the singleton instance of metricRecorder.
+// nil is returned if the recorder is not initialized.
+func Recorder() *metricRecorder {
+	return r
+}
+
+// NewRecorder initializes a new metricRecorder instance if it hasn't been initialized.
+func NewRecorder() *metricRecorder {
+	once.Do(func() {
+		r = &metricRecorder{
+			registry: metrics.NewKubeRegistry(),
+			metrics:  make(map[string]interface{}),
+		}
+	})
+	return r
+}
+
+// IncreaseCount increases the counter metric by 1.
+func (m *metricRecorder) IncreaseCount(name string, labels map[string]string) {
+	if m == nil {
+		return // recorder is not initialized
+	}
+
+	metric, ok := m.metrics[name]
+
+	if !ok {
+		klog.V(4).InfoS("Metric not found, registering", "name", name, "labels", labels)
+		m.registerCounterVec(name, "ebs_csi_aws_com metric", getLabelNames(labels))
+		m.IncreaseCount(name, labels)
+		return
+	}
+
+	metric.(*metrics.CounterVec).With(metrics.Labels(labels)).Inc()
+}
+
+// ObserveHistogram records the given value in the histogram metric.
+func (m *metricRecorder) ObserveHistogram(name string, value float64, labels map[string]string, buckets []float64) {
+	if m == nil {
+		return // recorder is not initialized
+	}
+	metric, ok := m.metrics[name]
+
+	if !ok {
+		klog.V(4).InfoS("Metric not found, registering", "name", name, "labels", labels, "buckets", buckets)
+		m.registerHistogramVec(name, "ebs_csi_aws_com metric", getLabelNames(labels), buckets)
+		m.ObserveHistogram(name, value, labels, buckets)
+		return
+	}
+
+	metric.(*metrics.HistogramVec).With(metrics.Labels(labels)).Observe(value)
+}
+
+func (m *metricRecorder) InitializeMetricsHandler(address, path string) {
+	if m == nil {
+		klog.InfoS("InitializeMetricsHandler: metric recorder is not initialized")
+		return
+	}
+
+	mux := http.NewServeMux()
+	mux.Handle(path, metrics.HandlerFor(
+		m.registry,
+		metrics.HandlerOpts{
+			ErrorHandling: metrics.ContinueOnError,
+		}))
+
+	go func() {
+		klog.InfoS("Metric server listening", "address", address, "path", path)
+
+		if err := http.ListenAndServe(address, mux); err != nil {
```

Comment:
Suggestion: harden security by adding a timeout. The default `http` package server comes with no timeout and is susceptible to DoS attacks.

```suggestion
		server := &http.Server{
				ReadTimeout: 3 * time.Second,
		}
		if err := server.ListenAndServe(address, mux); err != nil {
```

Reply:
✅